lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,258 @@
+// Adds the given note entries to the given ELF if the given ELF has a note section.
+// Otherwise, it does nothing.
+//
+// @param [in/out] elf : ELF to be updated with the new note entries.
+// @param notes        : An array of note entries to be inserted to the existing note section.
+void AddNotesToELF(SmallVectorImpl<char> &elf, llvm::ArrayRef<NoteEntry> notes) {
+  // Get ELF header that contains information for section header table offset
+  // and the number of section headers.
+  //
+  // Reference: http://www.skyfree.org/linux/references/ELF_Format.pdf
+  ELF::Elf64_Ehdr *ehdr = reinterpret_cast<ELF::Elf64_Ehdr *>(elf.data());
+
+  // Get the section headers and the existing note section.
+  MutableArrayRef<ELF::Elf64_Shdr> sectionHeaders(reinterpret_cast<ELF::Elf64_Shdr *>(elf.data() + ehdr->e_shoff),
+                                                  ehdr->e_shnum);
+  auto existingNoteSection = find_if(
+      sectionHeaders, [](const ELF::Elf64_Shdr &sectionHeader) { return sectionHeader.sh_type == ELF::SHT_NOTE; });
+
+  // We assume that the given ELF already contains a note section. Since AMD GPU
+  // accepts only ELFs with AMD related metadata, the assumption will be satisfied.
+  assert(existingNoteSection != sectionHeaders.end());
+
+  // Prepare the new note entries to be added to the existing note section.
+  const ELF::Elf64_Off newNoteEntryOffset = existingNoteSection->sh_offset + existingNoteSection->sh_size;
+  AppendingBinaryByteStream noteEntryStream(support::little);
+  prepareNoteEntries(notes, newNoteEntryOffset, noteEntryStream);
+
+  // Get the last section located just before the section header table.
+  auto sectionBeforeSectionHeaderTable =
+      std::max_element(sectionHeaders.begin(), sectionHeaders.end(),
+                       [ehdr](const ELF::Elf64_Shdr &largest, const ELF::Elf64_Shdr &current) {
+                         if (current.sh_offset > ehdr->e_shoff)
+                           return false;
+                         return current.sh_offset > largest.sh_offset;
+                       });
+
+  // Update the offset information of sections after the offset of new note entries.
+  // The new offset should be the offset where each section will be shifted to.
+  SmallVector<SectionShiftInfo> sectionAndNewOffset =
+      updateSectionOffsetsForShift(elf, newNoteEntryOffset, noteEntryStream.getLength(), sectionHeaders);
+
+  // Increase the size of the existing note section to include new note entries.
+  existingNoteSection->sh_size += noteEntryStream.getLength();
+
+  // Prepare the section header table shift if we have to shift it.
+  SmallVector<ELF::Elf64_Shdr> sectionHeaderTable;
+  if (ehdr->e_shoff > newNoteEntryOffset) {
+    ehdr->e_shoff = sectionBeforeSectionHeaderTable->sh_offset + sectionBeforeSectionHeaderTable->sh_size;
+    sectionHeaderTable.append(sectionHeaders.begin(), sectionHeaders.end());
```

Comment:
It took me a little bit of thinking to understand why you make a copy of `sectionHeaders`.  You might want to explain that with a comment.  It is very subtle, so I am not sure the best way to explain it.

Reply:
done